src/redux/note-content/reducers.ts

```diff
@@ -0,0 +1,127 @@
+} from './types'
+
+export const initialState: NoteContent = {
+  markdownContent: '',
+  id: '',
+  createTime: DateTime.fromSeconds(0),
+  lastChange: {
+    timestamp: DateTime.fromSeconds(0),
+    userId: ''
+  },
+  alias: '',
+  preVersionTwoNote: false,
+  viewCount: 0,
+  authorship: [],
+  noteTitle: '',
+  firstHeading: '',
+  metadata: {
+    title: '',
+    description: '',
+    tags: [],
+    deprecatedTagsSyntax: false,
+    robots: '',
+    lang: 'en',
+    dir: 'ltr',
+    breaks: true,
+    GA: '',
+    disqus: '',
+    type: '',
+    opengraph: new Map<string, string>()
+  }
+}
+
+export const NoteContentReducer: Reducer<NoteContent, NoteContentAction> = (state: NoteContent = initialState, action: NoteContentAction) => {
+  switch (action.type) {
+    case NoteContentActionType.SET_DOCUMENT_CONTENT:
+      return {
+        ...state,
+        markdownContent: (action as SetNoteContentAction).content
+      }
+    case NoteContentActionType.UPDATE_NOTE_TITLE_BY_FIRST_HEADING:
+      return {
+        ...state,
+        firstHeading: (action as UpdateNoteTitleByFirstHeadingAction).firstHeading,
+        noteTitle: generateNoteTitle(state.metadata, (action as UpdateNoteTitleByFirstHeadingAction).firstHeading)
+      }
+    case NoteContentActionType.SET_NOTE_DATA_FROM_SERVER:
+      return convertNoteToNoteContent((action as SetNoteDataFromServerAction).note)
+    case NoteContentActionType.SET_NOTE_META_DATA:
+      return {
+        ...state,
+        metadata: (action as SetNoteMetaDataFromRenderingAction).metadata,
+        noteTitle: generateNoteTitle((action as SetNoteMetaDataFromRenderingAction).metadata, state.firstHeading)
+      }
+    case NoteContentActionType.SET_CHECKBOX_IN_MARKDOWN_CONTENT:
+      return {
+        ...state,
+        markdownContent: setCheckboxInMarkdownContent(
+          state.markdownContent,
+          (action as SetCheckboxInMarkdownContentAction).lineInMarkdown,
+          (action as SetCheckboxInMarkdownContentAction).checked
+        )
+      }
+    default:
+      return state
+  }
+}
+
+const TASK_REGEX = /(\s*[-*] )(\[[ xX]])( .*)/
+const setCheckboxInMarkdownContent = (markdownContent: string, lineInMarkdown: number, checked: boolean): string => {
+  const lines = markdownContent.split('\n')
+  const results = TASK_REGEX.exec(lines[lineInMarkdown])
+  if (results) {
+    const before = results[1]
+    const after = results[3]
+    lines[lineInMarkdown] = `${before}[${checked ? 'x' : ' '}]${after}`
+    return lines.join('\n')
+  }
+  return markdownContent
+}
+
+const generateNoteTitle = (metaData: YAMLMetaData, firstHeading?: string) => {
+  if (metaData?.title && metaData?.title !== '') {
+    return metaData.title
+  } else if (metaData?.opengraph && metaData?.opengraph.get('title') && metaData?.opengraph.get('title') !== '') {
+    return (metaData?.opengraph.get('title') ?? firstHeading ?? '')
+  } else {
+    return (firstHeading ?? firstHeading ?? '').trim()
```

Comment:
We use a `.trim()` only in the last case. Wouldn't it make sense to use it on all execution branch returns of the method?

Reply:
That's right. Changed it.